src/agent/src/ccw.rs

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) IBM Corp. 2021
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::convert::TryInto;
+use std::fmt;
+use std::str::FromStr;
+
+use anyhow::anyhow;
+
+// CCW bus ID follow the format <xx>.<d>.<xxxx> [1, p. 11], where
+//   - <xx> is the channel subsystem ID, which is always 0 from the guest side, but different from
+//     the host side, e.g. 0xfe for virtio-*-ccw [1, p. 435],
+//   - <d> is the subchannel set ID, which ranges from 0-3 [2], and
+//   - <xxxx> is the device number (0000-ffff; leading zeroes can be omitted,
+//      e.g. 3 instead of 0003).
+// [1] https://www.ibm.com/docs/en/linuxonibm/pdf/lku4dd04.pdf
+// [2] https://qemu.readthedocs.io/en/latest/system/s390x/css.html
+
+// Maximum subchannel set ID
+const SUBCHANNEL_SET_MAX: u8 = 3;
+
+// CCW device. From the guest side, the first field is always 0 and can therefore be omitted.
+#[derive(Copy, Clone, Debug)]
+pub struct Device {
+    subchannel_set_id: u8,
+    device_number: u16,
+}
+
+impl Device {
+    pub fn new<T, U>(subchannel_set_id: T, device_number: U) -> anyhow::Result<Self>
+    where
+        T: fmt::Display + Copy + TryInto<u8>,
+        U: fmt::Display + Copy + TryInto<u16>,
+    {
+        let subchannel_set_id8: u8;
+        // TODO this can be prettified when https://github.com/rust-lang/rust/issues/53667 lands
+        let subchannel_err = Err(anyhow!(
+            "Subchannel set ID {} should be in range [0..{}]",
+            subchannel_set_id,
+            SUBCHANNEL_SET_MAX
+        ));
+        if let Ok(sid) = subchannel_set_id.try_into() {
+            if sid <= SUBCHANNEL_SET_MAX {
+                subchannel_set_id8 = sid;
+            } else {
+                return subchannel_err;
+            }
+        } else {
+            return subchannel_err;
+        }
+
+        if let Ok(device_number16) = device_number.try_into() {
+            return Ok(Device {
+                subchannel_set_id: subchannel_set_id8,
+                device_number: device_number16,
+            });
+        }
+        Err(anyhow!(
+            "Device number {} should be in range [{:04x}..{:04x}]",
+            device_number,
+            0,
+            u16::MAX
+        ))
+    }
+}
+
+impl FromStr for Device {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> anyhow::Result<Self> {
+        let split: Vec<&str> = s.split('.').collect();
+        if split.len() != 3 {
+            return Err(anyhow!(
+                "Wrong bus format. It needs to be in the form 0.<d>.<xxxx>, got {}",
```

Comment:
You might want to make that:

```rust
got {:?}
```

... so that if `s` is empty, that will be clear as the value will be quoted.